### PR TITLE
Add Pad CPU operator (including aligned padded shape support)

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
- It adds a new feature needed to implement splicing in RNN-T network.

#### What happened in this PR?

 - What solution was applied:
* Implement Pad CPU operator in terms of `SliceFlipNormalizePermutePad` kernel
* Adjust existing Pad GPU operator to use `SliceFlipNormalizePermutePad` kernel as well
* Remove redundant CPU pad kernel implementation
* Enhanced Pad operator API to support alignment on the padded shape

 - Affected modules and functionalities:
Pad operator

 - Key points relevant for the review:
`SliceFlipNormalizePermutePad` kernel args calculation

 - Validation and testing:
Existing C++ unit tests and python operator tests.
Added additional python tests to cover the aligned shape case

 - Documentation (including examples):
Updated operator documentation (docstr)

**JIRA TASK**: *[DALI-1236]*

### CODE REVIEW

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

File: dali/test/python/test_operator_pad.py [hunk 1]
```diff
@@ -70,12 +79,14 @@ def check_pad_synth_data(device, batch_size, input_max_shape, axes):
...
-    for device in ["gpu"]:
-        for batch_size in {1, 8, 100}:
-            for input_max_shape, axes in \
-                [((200,400,3), (0,)),
-                 ((200,400,3), (1,)),
-                 ((200,400,3), (0,1)),
-                 ((200,400,3), ()),
-                 ((200,400,3), [])]:
-                yield check_pad_synth_data, device, batch_size, input_max_shape, axes
+    for device in ["cpu"]:
```
Comment (JanuszL) [inline dali/test/python/test_operator_pad.py]:
```suggestion
    for device in ["cpu", "gpu"]:
```

File: dali/test/python/test_operator_pad.py [hunk 2]
```diff
@@ -70,12 +79,14 @@ def check_pad_synth_data(device, batch_size, input_max_shape, axes):
...
+        for batch_size in {1, 8}:
+            for input_max_shape, axes, align in \
+                [((200, 400, 3), (0,), None),
+                 ((200, 400, 3), (1,), None),
+                 ((200, 400, 3), (0, 1), None),
+                 ((200, 400, 3), (), None),
+                 ((200, 400, 3), [], None),
+                 ((200, 400, 3), (2,), (4,)),]:
+#                 ((200, 400, 3), (0, 1), (256, 256))]:
```
Comment (JanuszL) [inline dali/test/python/test_operator_pad.py]: ?

File: dali/test/python/test_operator_pad.py [hunk 3]
```diff
@@ -70,12 +79,14 @@ def check_pad_synth_data(device, batch_size, input_max_shape, axes):
...
-                yield check_pad_synth_data, device, batch_size, input_max_shape, axes
+    for device in ["cpu"]:
+        for batch_size in {1, 8}:
+            for input_max_shape, axes, align in \
+                [((200, 400, 3), (0,), None),
+                 ((200, 400, 3), (1,), None),
+                 ((200, 400, 3), (0, 1), None),
+                 ((200, 400, 3), (), None),
+                 ((200, 400, 3), [], None),
+                 ((200, 400, 3), (2,), (4,)),]:
```
Comment (JanuszL) [inline dali/test/python/test_operator_pad.py]: Can we test other alignments as well?

File: dali/test/python/test_operator_pad.py (hunk 1 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_pad.py]: I removed while debugging a specific case and forgot to revert. Thanks!

File: dali/test/python/test_operator_pad.py (hunk 2 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_pad.py]: Same as above, thanks!

File: dali/test/python/test_operator_pad.py (hunk 3 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_pad.py]: There are two cases:
```
                 ((200, 400, 3), (2,), (4,)),]:
                 ((200, 400, 3), (0, 1), (256, 256))]:
```
anything specific that you have in mind?

File: dali/kernels/slice/slice_flip_normalize_permute_cpu.h [hunk 4]
```diff
@@ -136,15 +141,16 @@ inline void SliceFlipNormalizePermuteImpl(OutputType *output, const InputType *i
...
+          std::integral_constant<int, DimsLeft - 1>());
       input += in_strides[d];
       output += out_strides[d];
     }
   }
 
   // zero pad
   for (; i < padded_out_shape[d]; i++) {
-    ZeroPad(output, out_strides, padded_out_shape,
+    ZeroPad(output, out_strides, padded_out_shape, fill_value,
```
Comment (mzient) [inline dali/kernels/slice/slice_flip_normalize_permute_cpu.h]: It's no longer _Zero_ pad, is it?

Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_permute_cpu.h]: Correct, I will fix it

File: dali/kernels/slice/slice_flip_normalize_permute_cpu.h [hunk 5]
```diff
@@ -61,11 +61,12 @@ template <typename OutputType, int Dims>
 inline void ZeroPad(OutputType *output,
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_permute_cpu.h]: Sure, will do

File: dali/kernels/slice/slice_flip_normalize_permute_cpu.h [hunk 6]
```diff
@@ -58,28 +58,30 @@ struct NormalizePolicy {
...
+inline void Pad(OutputType *output,
+                TensorShape<Dims> out_strides,
+                TensorShape<Dims> padded_out_shape,
+                float fill_value,
+                std::integral_constant<int, 1>) {
   constexpr auto d = Dims - 1;
   // zero pad
   for (int64_t i = 0; i < padded_out_shape[d]; i++) {
-    *output = 0;
+    *output = fill_value;
```
Comment (banasraf) [inline dali/kernels/slice/slice_flip_normalize_permute_cpu.h]:
```suggestion
   *output = static_cast<OutputType>(fill_value);
```
just like it is done below?

File: dali/kernels/slice/slice_flip_normalize_permute_cpu.h [hunk 7]
```diff
@@ -136,16 +141,17 @@ inline void SliceFlipNormalizePermuteImpl(OutputType *output, const InputType *i
...
           output, input, in_strides, out_strides, out_shape, padded_out_shape, mean + norm_idx,
-          inv_stddev + norm_idx, normalization_dim, std::integral_constant<int, DimsLeft - 1>());
+          inv_stddev + norm_idx, normalization_dim, fill_value,
+          std::integral_constant<int, DimsLeft - 1>());
       input += in_strides[d];
       output += out_strides[d];
     }
   }
 
   // zero pad
```
Comment (banasraf) [inline dali/kernels/slice/slice_flip_normalize_permute_cpu.h]:
```suggestion
  // pad
```

File: dali/operators/util/pad.cc [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  output_desc.resize(1);
+  const auto &input = ws.template InputRef<CPUBackend>(0);
+  auto in_shape = input.shape();
+  int ndim = input.shape().sample_dim();
+  int nsamples = input.size();
+  auto nthreads = ws.GetThreadPool().size();
+
+  TYPE_SWITCH(input.type().id(), type2id, T, PAD_SUPPORTED_TYPES, (
+    VALUE_SWITCH(ndim, Dims, PAD_SUPPORTED_NDIMS, (
+      using Kernel = kernels::SliceFlipNormalizePermuteCPU<T, T, Dims>;
```
Comment (JanuszL) [inline dali/operators/util/pad.cc]:
```suggestion
      using Kernel = kernels::SliceFlipNormalizePermutePadCPU<T, T, Dims>;
```

File: dali/kernels/slice/slice_flip_normalize_permute_cpu.h (hunk 6 above)
Comment (mzient) [reply dali/kernels/slice/slice_flip_normalize_permute_cpu.h]: ...why not Convert<OutputType>?

File: dali/operators/util/pad.cc (hunk 8 above)
Comment (jantonguirao) [reply dali/operators/util/pad.cc]: changed

File: dali/kernels/slice/slice_flip_normalize_permute_cpu.h (hunk 7 above)
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_permute_cpu.h]: done

File: dali/kernels/slice/slice_flip_normalize_permute_cpu.h (hunk 6 above)
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_permute_cpu.h]: done

File: dali/test/python/test_operator_pad.py (hunk 3 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_pad.py]: added two more cases

File: dali/operators/util/pad.cu [hunk 9]
```diff
@@ -17,99 +17,60 @@
...
 
-      output_desc[0].type = TypeInfo::Create<DataType>();
-      output_desc[0].shape.resize(batch_size_, NumAxes);
-      kmgr_.Initialize<Kernel>();
+      auto in_view = view<const T, Dims>(input);
+      auto &kernel_sample_args = FillArgs<Args>(in_shape, in_layout);
 
-      auto in_view = view<const DataType, NumAxes>(input);
+      kmgr_.Initialize<Kernel>();
+      kmgr_.Resize<Kernel>(1, 1);
```
Comment (mzient) [inline dali/operators/util/pad.cu]: The call to `Initialize` is pointless - the purpose of this function is to populate the slots allocated by Resize without Kernel template argument. When using typed Resize, Initialize is redundant - and in any case, should be called _after_ Resize.

File: dali/operators/util/pad.cc [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+  int nsamples = input.size();
+  auto nthreads = ws.GetThreadPool().size();
+
+  TYPE_SWITCH(input.type().id(), type2id, T, PAD_SUPPORTED_TYPES, (
+    VALUE_SWITCH(ndim, Dims, PAD_SUPPORTED_NDIMS, (
+      using Kernel = kernels::SliceFlipNormalizePermutePadCpu<T, T, Dims>;
+      using Args = kernels::SliceFlipNormalizePermutePadArgs<Dims>;
+
+      kmgr_.Initialize<Kernel>();
+      kmgr_.Resize<Kernel>(nthreads, nsamples);
```
Comment (mzient) [inline dali/operators/util/pad.cc]: Same as in GPU - redundant Initialize.

File: dali/operators/util/pad.cu [hunk 11]
```diff
@@ -17,99 +17,60 @@
...
+      auto req = kmgr_.Setup<Kernel>(0, ctx, in_view, kernel_sample_args);
 
-      auto &req = kmgr_.Setup<Kernel>(0, ctx, in_view, axes_);
+      output_desc[0].type = TypeInfo::Create<T>();
+      output_desc[0].shape.resize(batch_size_, Dims);
```
Comment (mzient) [inline dali/operators/util/pad.cu]: Use `input.shape.num_samples()` instead?

File: dali/operators/util/pad.cc [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+      output_desc[0].type = TypeInfo::Create<T>();
+      output_desc[0].shape.resize(batch_size_, Dims);
```
Comment (mzient) [inline dali/operators/util/pad.cc]:
```suggestion
      output_desc[0].shape.resize(nsamples, Dims);
```

File: dali/operators/util/pad.cc [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+
+      auto in_view = view<const T, Dims>(input);
+      auto &kernel_sample_args = FillArgs<Args>(in_shape, in_layout);
+      for (int i = 0; i < batch_size_; i++) {
```
Comment (mzient) [inline dali/operators/util/pad.cc]:
```suggestion
      for (int i = 0; i < nsamples; i++) {
```

File: dali/operators/util/pad.cc [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+            [3, 199,  5, -1, -1]]
+
+- `1-D` samples, `fill_value` = -1, `axes` = (0,), `shape` = (7,)
+
+::
+
+  input  = [[3,   4, 2, 5, 4],
+            [2,   2],
+            [3, 199, 5]];
+  output = [[3, 4,   2,  5,  4, -1, -1],
```
Comment (mzient) [inline dali/operators/util/pad.cc]: Nitpick: inconsistent alignment.

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Pull request merged**